File: quant-black-box/backend/app/routers/heston.py
from fastapi import APIRouter

from .. import engine
from ..schemas import (
    GridPoint,
    GridResponse,
    HestonParams as HestonParamsModel,
    HestonRequest,
    HestonResponse,
)

router = APIRouter()


def build_grid(req: HestonRequest) -> GridResponse:
    n = req.grid.n
    x0, x1 = req.K * 0.4, req.K * 1.6
    y0, y1 = 0.05, 1.6
    params = engine.HestonParams(
        r=req.r, v0=req.v0, kappa=req.kappa, theta=req.theta, sigv=req.sigv, rho=req.rho
    )
    points: list[GridPoint] = []
    for i in range(n):
        x = x0 + (x1 - x0) * i / (n - 1)
        for j in range(n):
            t = y0 + (y1 - y0) * j / (n - 1)
            if req.grid.metric == "delta":
                d = engine.heston_p1(x, req.K, t, params)
                z = d if req.opt == "call" else d - 1.0
            else:
                mm = engine.heston_price(x, req.K, t, params)
                z = mm.C if req.opt == "call" else mm.P
            points.append(GridPoint(x=x, y=t, z=z))
    return GridResponse(
        n=n,
        metric=req.grid.metric,
        points=points,
        shape=[n, n],
        x_range=[x0, x1],
        y_range=[y0, y1],
    )


@router.post("", response_model=HestonResponse)
def price(req: HestonRequest) -> HestonResponse:
    params = engine.HestonParams(
        r=req.r, v0=req.v0, kappa=req.kappa, theta=req.theta, sigv=req.sigv, rho=req.rho
    )
    m = engine.heston_price(req.S0, req.K, req.T, params)
    return HestonResponse(
        opt=req.opt,
        params=HestonParamsModel(
            S0=req.S0, K=req.K, T=req.T, r=req.r,
            v0=req.v0, kappa=req.kappa, theta=req.theta, sigv=req.sigv, rho=req.rho,
        ),
        C=m.C,
        P=m.P,
        delta=m.delta,
        deltaP=m.deltaP,
        grid=build_grid(req) if req.grid is not None else None,
    )

File: quant-black-box/server/src/routes/heston.ts
import { Router } from 'express';
import { z } from 'zod';
import { hestonPrice, hestonP1 } from '../engine.js';

const router = Router();

const gridSchema = z.object({
  n: z.coerce.number().min(2).max(120).default(26),
  metric: z.enum(['price', 'delta']).default('price'),
});

const hestonSchema = z.object({
  S0: z.coerce.number().gt(0),
  K: z.coerce.number().gt(0),
  T: z.coerce.number().gt(0),
  r: z.coerce.number(),
  v0: z.coerce.number().gte(0),
  kappa: z.coerce.number().gte(0),
  theta: z.coerce.number().gte(0),
  sigv: z.coerce.number().gte(0),
  rho: z.coerce.number().min(-1).max(1),
  opt: z.enum(['call', 'put']).default('call'),
  grid: gridSchema.nullish(),
});

router.post('/', (req, res) => {
  const parsed = hestonSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(422).json({ ok: false, error: 'Invalid parameters', details: parsed.error.flatten() });
    return;
  }
  const { S0, K, T, r, v0, kappa, theta, sigv, rho, opt, grid } = parsed.data;
  const params = { r, v0, kappa, theta, sigv, rho };
  const m = hestonPrice(S0, K, T, params);

  let gridData = null;
  if (grid) {
    const n = grid.n;
    const x0 = K * 0.4, x1 = K * 1.6, y0 = 0.05, y1 = 1.6;
    const points: { x: number; y: number; z: number }[] = [];
    for (let i = 0; i < n; i++) {
      const x = x0 + (x1 - x0) * i / (n - 1);
      for (let j = 0; j < n; j++) {
        const t = y0 + (y1 - y0) * j / (n - 1);
        let z: number;
        if (grid.metric === 'delta') {
          const d = hestonP1(x, K, t, params);
          z = opt === 'call' ? d : d - 1;
        } else {
          const mm = hestonPrice(x, K, t, params);
          z = opt === 'call' ? mm.C : mm.P;
        }
        points.push({ x, y: t, z });
      }
    }
    gridData = { n, metric: grid.metric, points, shape: [n, n], x_range: [x0, x1], y_range: [y0, y1] };
  }

  res.json({
    opt,
    params: { S0, K, T, r, v0, kappa, theta, sigv, rho },
    C: m.C, P: m.P, delta: m.delta, deltaP: m.deltaP,
    grid: gridData,
  });
});

export default router;
